dialogue-core/src/main/java/com/palantir/dialogue/core/Balanced.java

```diff
@@ -0,0 +1,172 @@
+
+    /**
+     * We'd like to remember failures for a long time, but this would increase CPU on a hot codepath as we compute
+     * the number of failures for each channel in order to sort. See {@link ChannelWithStats#score}.
+     */
+    private static final Duration FAILURE_MEMORY = Duration.ofSeconds(5);
+
+    /**
+     * This comparator is a little risky because the data can change while we're sorting. In practise the sort
+     * finishes a lot quicker than network requests do, and we don't mind if it's not exactly right.
+     */
+    @VisibleForTesting
+    static final Comparator<ChannelWithStats> BY_SCORE = Comparator.comparing(ChannelWithStats::score);
+
+    private final ImmutableList<ChannelWithStats> channels;
+    private final Random random;
+    private final CodahaleClock clock;
+
+    Balanced(ImmutableList<LimitedChannel> channels, Random random, Ticker ticker) {
+        Preconditions.checkState(channels.size() >= 2, "At least two channels required");
+        this.random = random;
+        this.clock = new CodahaleClock(ticker);
+        this.channels = channels.stream()
+                .map(channel -> new ChannelWithStats(channel, clock))
+                .collect(ImmutableList.toImmutableList());
+
+        log.debug("Initialized", SafeArg.of("count", channels.size()), UnsafeArg.of("channels", channels));
+    }
+
+    @Override
+    public Optional<ListenableFuture<Response>> maybeExecute(Endpoint endpoint, Request request) {
+        if (log.isDebugEnabled()) {
+            log.debug(
+                    "Channel scores {}",
+                    SafeArg.of("scores", Lists.transform(channels, ChannelWithStats::score)),
+                    SafeArg.of("inflight", Lists.transform(channels, ChannelWithStats::inflight)));
+        }
+
+        // pre-shuffling is pretty important here, otherwise when there are no requests in flight, we'd
+        // *always* prefer the first channel of the list, leading to a higher overall load.
+        List<ChannelWithStats> preShuffled = shuffleImmutableList(channels, random);
+
+        // TODO(dfox): P2C optimization when we have high number of nodes to save CPU?
+        // http://www.eecs.harvard.edu/~michaelm/NEWWORK/postscripts/twosurvey.pdf
+        return preShuffled.stream()
+                .sorted(BY_SCORE)
+                .map(channel -> channel.maybeExecute(endpoint, request))
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                // we rely heavily on streams being lazy here
+                .findFirst();
+    }
+
+    /** Returns a new shuffled list, without mutating the input list (which may be immutable). */
+    private static <T> List<T> shuffleImmutableList(ImmutableList<T> sourceList, Random random) {
+        List<T> mutableList = new ArrayList<>(sourceList);
+        Collections.shuffle(mutableList, random);
+        return mutableList;
+    }
```

Comment:
I don't think the shuffle buys us anything because utilization is immediately reflected. We shouldn't need to break ties because we have enough data that either they're uncommon, or it doesn't matter which channel we select because we know it works just as well.

Reply:
Are you sure? if we have hundreds of clients and none of them do this shuffle, then every single time there's a client has a blip of zero active requests, they will always pick the first channel to hit first. The simulations show a pretty noticeable effect where server utilization is just not balanced anymore.

Reply:
That’s a good point, I suppose we could do something similar to the random selection implementation

Reply:
Given that the number of nodes is pretty much always <10, I think it might be fine to leave it as is for now tbh